File: src/cli/env-file.js
import { readFileSync, writeFileSync, existsSync } from "node:fs";

const LINE_RE = /^([A-Z_][A-Z0-9_]*)=(.*)$/;

export function readEnvFile(filePath) {
  if (!existsSync(filePath)) return {};
  const content = readFileSync(filePath, "utf8");
  const records = {};
  for (const line of content.split("\n")) {
    const m = line.match(LINE_RE);
    if (m) {
      records[m[1]] = m[2];
    }
  }
  return records;
}

export function writeEnvFile(filePath, records) {
  const lines = Object.entries(records).map(([k, v]) => `${k}=${v}`);
  writeFileSync(filePath, lines.join("\n") + "\n", "utf8");
}

export function updateEnvValue(filePath, key, value) {
  const records = readEnvFile(filePath);
  if (value === "" || value === undefined || value === null) {
    delete records[key];
  } else {
    records[key] = value;
  }
  writeEnvFile(filePath, records);
  return records;
}

const SECRET_PATTERNS = [/KEY/i, /SECRET/i, /TOKEN/i, /PASSWORD/i];

export function isSecretKey(key) {
  return SECRET_PATTERNS.some((p) => p.test(key));
}

export function maskValue(val) {
  if (!val) return "";
  if (val.length <= 8) return "***";
  return val.slice(0, 3) + "***" + val.slice(-4);
}

File: atlas-py/atlas_py/cli/env_file.py
"""Helpers for reading and writing .env files. Mirrors src/cli/env-file.js."""

from __future__ import annotations

import os
import re

_LINE_RE = re.compile(r"^([A-Z_][A-Z0-9_]*)=(.*)$")

SECRET_PATTERNS = [re.compile(p) for p in (r"KEY", r"SECRET", r"TOKEN", r"PASSWORD")]


def read_env_file(file_path):
    if not os.path.exists(file_path):
        return {}
    with open(file_path, "r", encoding="utf8") as f:
        content = f.read()
    records = {}
    for line in content.split("\n"):
        m = _LINE_RE.match(line)
        if m:
            records[m.group(1)] = m.group(2)
    return records


def write_env_file(file_path, records):
    lines = [f"{k}={v}" for k, v in records.items()]
    with open(file_path, "w", encoding="utf8") as f:
        f.write("\n".join(lines) + "\n")


def update_env_value(file_path, key, value):
    records = read_env_file(file_path)
    if value == "" or value is None:
        records.pop(key, None)
    else:
        records[key] = value
    write_env_file(file_path, records)
    return records


def is_secret_key(key):
    return any(p.search(key) for p in SECRET_PATTERNS)


def mask_value(val):
    if not val:
        return ""
    if len(val) <= 8:
        return "***"
    return val[:3] + "***" + val[-4:]
